mono/utils/mono-proclib.c

```diff
@@ -838,15 +838,24 @@ mono_cpu_count (void)
  * [5] https://github.com/dotnet/coreclr/blob/7058273693db2555f127ce16e6b0c5b40fb04867/src/pal/src/misc/sysinfo.cpp#L148
  */
 
+	int count = 0;
```

Comment:
Have you looked at how `mono_cpu_count` gets used? In some cases, it may need the total processor count.

coreclr has two separate functions `TotalProcessorCount` and `CurrentProcessCpuCount`. The latter takes into account the cgroup limit, the former does not.

Reply:
Mono grabs the count in these circumstances:
#### mono/profiler/log-args.c
* load_args_from_env_or_default: `config->max_allocated_sample_hits = mono_cpu_count () * 1000;`
#### mono/metadata/threadpool-worker-default.c
* mono_threadpool_worker_init: `threads_count = mono_cpu_count () * threads_per_cpu;`
* mono_threadpool_worker_set_max: `cpu_count = mono_cpu_count ();`
#### mono/metadata/icall.c 
* ves_icall_System_Environment_get_ProcessorCount: `return mono_cpu_count ();` 
#### mono/metadata/mono-perfcounters.c 
* get_cpu_instances:`count = mono_cpu_count () + 1; /* +1 for "_Total" */`
#### mono/metadata/threadpool.c
* initialize: `threadpool.limit_io_min = mono_cpu_count ();`
* ves_icall_System_Threading_ThreadPool_SetMaxThreadsNative: `gint cpu_count = mono_cpu_count ();`
#### metadata/icall-eventpipe.c
* mono_eventpipe_init: `table->ep_rt_mono_cpu_count = mono_cpu_count;`
#### mono/sgen/sgen-marksweep.c
* sgen_marksweep_init_internal: 
  * `if (mono_cpu_count () <= 1) is_parallel = FALSE;`
  * `sgen_workers_create_context (GENERATION_OLD, mono_cpu_count ());`
#### mono/sgen/sgen-simple-nursery.c
* sgen_simple_nursery_init: 
  * `if (mono_cpu_count () <= 1) parallel = FALSE;`
  * `sgen_workers_create_context (GENERATION_NURSERY, mono_cpu_count ());`
#### mono/mini/mini-ppc.c
* mono_arch_init: `if (mono_cpu_count () > 1) cpu_hw_caps |= PPC_SMP_CAPABLE;`
#### mono/utils/mono-proclib.c
* get_cpu_times: `uhz *= mono_cpu_count ();`
* mono_cpu_usage: `cpu_total_time = (current_time - (prev ? prev->current_time : 0)) * mono_cpu_count ();`
#### mono-proclib-windows.c
* Defines its own version of `mono_cpu_count`:
```
int
mono_cpu_count (void)
{
        SYSTEM_INFO info;
        GetSystemInfo (&info);
        return info.dwNumberOfProcessors;
}
```
* mono_cpu_usage: `guint64 total_available_time = (current_time - creation_time) * mono_cpu_count ();`

### Questions

The basic question is "what is the appropriate cpu count for each of these cases?". From which other questions are generated. For example,
* What does it mean to measure performance in a containerized environment? Are you measuring what it perceives its environment to be or what performance its getting from its underlying system? 
* For sgen determining if it's a parallel environment is it `min(phys, restricted) > 1`?
* What are the applications that end up invoking `ves_icall_System_Environment_get_ProcessorCount()` expecting to see?
* What needs to be done with `mono-proclib-windows.c`?

I will attempt to fill in this table and welcome/encourage advice. Hybrid means some algorithm based on the real and restricted values (min, max ...).

| Function | Real  | Restricted | Hybrid |
| -------- | ------| --------- | ------ |
| `load_args_from_env_or_default` | | |
| `mono_threadpool_worker_init` | | |
| `mono_threadpool_worker_set_max` | | |
| `ves_icall_System_Environment_get_ProcessorCount` | | | 
| `get_cpu_instances` | | |
| `initialize` | | |
| `ves_icall_System_Threading_ThreadPool_SetMaxThreadsNative` | | |
| `mono_eventpipe_init` | | |
| `sgen_marksweep_init_internal` | | | 
| `sgen_simple_nursery_init` | | |
| `mono_arch_init` | | |
| `get_cpu_times` | | |
| `mono_cpu_usage` | | |

Reply:
First of all, the "restricted" CPU count should always be bounded by the "real" CPU count, right from the initial computation.   Apart from that, I don't think there's any other "hybrid" computation needed.

Now, as to when the "real" vs. "restricted" CPU count is appropriate, I think the following principles apply:

- The number provided to managed code via `System.Environment.GetProcessorCount` should be the restricted count, to match what CoreCLR is doing.
- Whenever we create a pool of worker threads (whether for internal work like GC or user work), the number of workers should use the restricted count - using more will likely not help but rather overload the machine.
- When determining whether the system is single-CPU (e.g. to avoid using locking primitives), we must use the real CPU count.
- When handling per-CPU data like performance counters or samples, those need to use the real CPU count.
- I think we shouldn't touch Windows for now - CGroup is a Linux-specific concept, and if there's anything similar to be done for Windows that can be done in a separate patch by someone who actually understands Windows.

This should cover all the above use case, except for `ep_rt_mono_cpu_count`.  I don't actually understand how this works in the Mono sources, I see no user (or even definition).   In the dotnet runtime eventpipe code, there's instead a routine `ep_rt_processors_get_count`, which currently uses `mono_cpu_count ()`.  The equivalent CoreCLR implementation uses `GetSystemInfo() .dwNumberOfProcessors`, which uses the "real" count.

So overall I would complete the above table as follows:

Function | Real | Restricted
-- | -- | -- 
`load_args_from_env_or_default` | per-CPU data  |    
`mono_threadpool_worker_init` | | thread pool    
`mono_threadpool_worker_set_max` | | thread pool   
`ves_icall_System_Environment_get_ProcessorCount` | | API compat
`get_cpu_instances` |  per-CPU data | 
`initialize` | | thread pool  
`ves_icall_System_Threading_ThreadPool_SetMaxThreadsNative` | | thread pool  
`mono_eventpipe_init` | see above  |  
`sgen_marksweep_init_internal` |   | thread pool 
`sgen_simple_nursery_init` |   | thread pool 
`mono_arch_init` |  SMP flag  |  
`get_cpu_times` | per-CPU data  |   
`mono_cpu_usage` |  | used for thread pool

Reply:
Changes made and pushed